addons/manager.go

```diff
@@ -0,0 +1,282 @@
+	}
+	return clientconfig, nil
+}
+
+func runHubManager(ctx context.Context, options AddonAgentOptions) {
+	hubConfig, err := GetClientConfig(options.HubKubeconfigFile)
+	if err != nil {
+		klog.Error(err, "unable to get kubeconfig")
+		os.Exit(1)
+	}
+
+	spokeClient, err := GetClientFromConfig(ctrl.GetConfigOrDie())
+	if err != nil {
+		klog.Error(err, "unable to get spoke client")
+		os.Exit(1)
+	}
+
+	mgr, err := ctrl.NewManager(hubConfig, ctrl.Options{
+		Scheme:                 mgrScheme,
+		Port:                   9443,
+		MetricsBindAddress:     "0", // disable metrics
+		HealthProbeBindAddress: "0", // disable health probe
+		ReadinessEndpointName:  "0", // disable readiness probe
+		Namespace:              options.SpokeClusterName,
+	})
+	if err != nil {
+		klog.Error(err, "unable to start manager")
+		os.Exit(1)
+	}
+
+	if err = (&MirrorPeerReconciler{
+		Scheme:           mgr.GetScheme(),
+		HubClient:        mgr.GetClient(),
+		SpokeClient:      spokeClient,
+		SpokeClusterName: options.SpokeClusterName,
+	}).SetupWithManager(mgr); err != nil {
+		klog.Error(err, "unable to create controller", "controller", "MirrorPeer")
+		os.Exit(1)
+	}
+
+	if err = (&GreenSecretReconciler{
+		Scheme:           mgr.GetScheme(),
+		HubClient:        mgr.GetClient(),
+		SpokeClient:      spokeClient,
+		SpokeClusterName: options.SpokeClusterName,
+	}).SetupWithManager(mgr); err != nil {
+		klog.Error(err, "unable to create controller", "controller", "GreenSecret")
+		os.Exit(1)
+	}
+
+	klog.Info("starting hub controller manager")
+	if err := mgr.Start(ctx); err != nil {
+		klog.Error(err, "problem running hub controller manager")
+		os.Exit(1)
+	}
+}
+
+func runSpokeManager(ctx context.Context, options AddonAgentOptions) {
+	hubConfig, err := GetClientConfig(options.HubKubeconfigFile)
+	if err != nil {
+		klog.Error(err, "unable to get kubeconfig")
+		os.Exit(1)
+	}
+
+	hubClient, err := GetClientFromConfig(hubConfig)
+	if err != nil {
+		klog.Error(err, "unable to get spoke client")
+		os.Exit(1)
+	}
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:                 mgrScheme,
+		Port:                   9443,
+		MetricsBindAddress:     "0", // disable metrics
+		HealthProbeBindAddress: "0", // disable health probe
+		ReadinessEndpointName:  "0", // disable readiness probe
+	})
+
+	if err != nil {
+		klog.Error(err, "unable to start manager")
+		os.Exit(1)
+	}
+
+	if err = mgr.Add(manager.RunnableFunc(func(ctx context.Context) error {
+		klog.Infof("Waiting for MaintenanceMode CRD to be created. MaintenanceMode controller is not running yet.")
+		// Wait for 45s as it takes time for MaintenanceMode CRD to be created.
+		return runtimewait.PollUntilWithContext(ctx, 15*time.Second,
+			func(ctx context.Context) (done bool, err error) {
+				var crd extv1.CustomResourceDefinition
+				readErr := mgr.GetAPIReader().Get(ctx, types.NamespacedName{Name: "maintenancemodes.ramendr.openshift.io"}, &crd)
```

Comment:
maintenancemodes.ramendr.openshift.io => Use constant or contruct from the ramen api itself.